numpyro/contrib/module.py

```diff
@@ -71,6 +82,150 @@ def haiku_module(name, nn, input_shape=None):
             raise ValueError('Valid value for `input_shape` needed to initialize.')
         # feed in dummy data to init params
         rng_key = numpyro.sample(name + '$rng_key', PRNGIdentity())
-        nn_params = nn.init(rng_key, jnp.ones(input_shape))
-        numpyro.param(module_key, nn_params)
-    return partial(nn.apply, nn_params, None)
+        nn_params = nn_module.init(rng_key, jnp.ones(input_shape))
+        # haiku init returns an immutable dict
+        # we cast it to a mutable one to be able to set priors for parameters
+        nn_params = numpyro.param(module_key, haiku.data_structures.to_mutable_dict(nn_params))
+    return partial(nn_module.apply, nn_params, None)
+
+
+# register an "empty" parameter which only stores its shape
+# so that the optimizer can skip optimize this parameter, while
+# it still provides shape information for priors
+ParamShape = namedtuple("ParamShape", ["shape"])
+register_pytree_node(ParamShape, lambda x: ((None,), x.shape), lambda shape, x: ParamShape(shape))
+
+
+def _update_params(params, new_params, prior, prefix=''):
+    """
+    A helper to recursively set prior to new_params.
+    """
+    for name, item in params.items():
+        flatten_name = ".".join([prefix, name]) if prefix else name
+        if isinstance(item, dict):
+            assert not isinstance(prior, dict) or flatten_name not in prior
+            new_item = new_params[name]
+            _update_params(item, new_item, prior, prefix=flatten_name)
+        elif (not isinstance(prior, dict)) or flatten_name in prior:
+            d = prior[flatten_name] if isinstance(prior, dict) else prior
+            if isinstance(params[name], ParamShape):
+                param_shape = params[name].shape
+            else:
+                param_shape = jnp.shape(params[name])
+                params[name] = ParamShape(param_shape)
+            param_batch_shape = param_shape[:len(param_shape) - d.event_dim]
+            # XXX: here we set all dimensions of prior to event dimensions.
+            new_params[name] = numpyro.sample(flatten_name, d.expand(param_batch_shape).to_event())
+
+
+def random_flax_module(name, nn_module, prior, *, input_shape=None):
+    """
+    A primitive to place a prior over the parameters of the Flax module `nn_module`.
+
+    :param str name: name of NumPyro module
+    :param flax.nn.Module: the module to be registered with NumPyro
+    :param prior: a NumPyro distribution or a Python dict with parameter names as keys and
+        respective distributions as values. For example::
+
+            net = random_flax_module("net",
+                                     flax.nn.Dense.partial(features=1),
+                                     prior={"bias": dist.Cauchy(), "kernel": dist.Normal()},
```

Comment:
For an MLP, if we just want to make the last layer bayesian, would this API allow for it and how would we do that (would a delta distribution work here or can we just skip some parameters entirely)?

Reply:
Yes, `update_params` only lifts those in `prior` to `sample` statements (see its test, where the parameters `a.b.e` is kept as-is). At `net$params` site:
+ the nested structure is kept as-is
+ parameters not in `prior` are kept as-is
+ parameters in `prior` becomes `ParamShape` (which only holds `shape` information). When the optimizer flattens those values, they become empty. So we don't optimize anything here.

The output of this random module will be an NN with the following parameters (stored in `new_params` in the code):
+ the nested structure is kept as-is
+ parameters not in `prior` are kept as-is
+ parameters in `prior` will take the values drawn from their corresponding prior.

So the make the last layer bayesian, we can set `prior={"Dense_3.kernel": dist.Normal(0, 1), "Dense_3.bias": dist.Normal(0, 1)}`. Because I used flatten-with-dot name convention in `prior`, I think we can make `_update_params` more flexibile by using regex, i.e. `prior={"Dense_3.*": dist.Normal(0, 1)}`. I can do it in the follow-up PR (I believe it just took two or three lines of code to replace the check `flatten_name in prior` by a pattern matching for the regex expressions in `prior`'s keys). What do you think?

(using regex can be helpful when we want to set different priors for `kernel` and `bias`, across all layers)

Reply:
You are right. I also find it tricky to specify correct path for parameters. This might be not a problem of flax/haiku users but for new users who want to construct a bnn, it is inconvenient. Looking like https://github.com/google/objax is the solution for us. Its interface is pretty similar to PyTorch.